### ISSUE TITLE
`iptables` command does not work in k8s.gcr.io/build-image/debian-iptables-amd64:v11.1.0

### ISSUE THREAD

**Issue by listx** [kubernetes/release#1594]
#### What happened: 

Filing this bug with info passed on by Taahir Ahmed, a fellow Googler.

Ran `iptables` but it failed with an error message.

```
taahm@taahm:~$ kubectl run -it --generator=run-pod/v1 --image=k8s.gcr.io/build-image/debian-iptables-amd64:v11.1.0 iptables-pod
If you don't see a command prompt, try pressing enter.

# iptables
update-alternatives: error: alternative /usr/sbin/iptables-legacy for iptables not registered; not setting
```

#### What you expected to happen:

Run without error.

#### How to reproduce it (as minimally and precisely as possible):

See "What happened" section.

#### Anything else we need to know?:

Taahir says:
> That means that the iptables binary on the system is a little shim that just prints out an error message, and we need to run the update-alternatives tool to select a provider. I think we are getting caught be the Debian iptables to nftables migration documented here: https://wiki.debian.org/iptables

A new v11.1.1 version that fixes the problem should suffice.

#### Environment:

- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Docker image: `k8s.gcr.io/build-image/debian-iptables-amd64:v11.1.0`

**Comment by wespanther** [comment]
@justaugustus suggested it might be how we're invoking this above, pointed me to this script:

https://github.com/kubernetes/release/blob/master/images/build/debian-iptables/stretch/iptables-wrapper

I’m seeing the same thing just pulling the docker file and executing iptables, it’s definitely using the wrapper script:

```
wpanther@wpanther /usr/local/google/home/wpanther » docker run -it k8s.gcr.io/build-image/debian-iptables-amd64:v11.1.0 /usr/sbin/iptables
update-alternatives: error: alternative /usr/sbin/iptables-legacy for iptables not registered; not setting
```

**Comment by wespanther** [comment]
So I think the gist of the problem is that iptables is being replaced by nftables starting in buster. You can see iptables v1.8.3 provides xtables-legacy-multi and xtables-nft-multi, and presumably sets up those symlinks (I haven't looked at the source yet). 

```
wpanther@wpanther /usr/local/google/home/wpanther » docker run -it k8s.gcr.io/build-image/debian-iptables-amd64:v12.1.0 sh
# /sbin/iptables --version
iptables v1.8.3 (legacy)
# ls -al /usr/sbin/iptables-*
-rwxr-xr-x 1 root root 7057 Jul 17  2019 /usr/sbin/iptables-apply
lrwxrwxrwx 1 root root   20 Jul 17  2019 /usr/sbin/iptables-legacy -> xtables-legacy-multi
lrwxrwxrwx 1 root root   20 Jul 17  2019 /usr/sbin/iptables-legacy-restore -> xtables-legacy-multi
lrwxrwxrwx 1 root root   20 Jul 17  2019 /usr/sbin/iptables-legacy-save -> xtables-legacy-multi
lrwxrwxrwx 1 root root   17 Jul 17  2019 /usr/sbin/iptables-nft -> xtables-nft-multi
lrwxrwxrwx 1 root root   17 Jul 17  2019 /usr/sbin/iptables-nft-restore -> xtables-nft-multi
lrwxrwxrwx 1 root root   17 Jul 17  2019 /usr/sbin/iptables-nft-save -> xtables-nft-multi
lrwxrwxrwx 1 root root   34 May  6 01:05 /usr/sbin/iptables-restore -> /etc/alternatives/iptables-restore
lrwxrwxrwx 1 root root   17 Jul 17  2019 /usr/sbin/iptables-restore-translate -> xtables-nft-multi
lrwxrwxrwx 1 root root   31 May  6 01:05 /usr/sbin/iptables-save -> /etc/alternatives/iptables-save
lrwxrwxrwx 1 root root   17 Jul 17  2019 /usr/sbin/iptables-translate -> xtables-nft-multi
-rwxr-xr-x 1 root root 1684 May  6 01:05 /usr/sbin/iptables-wrapper
# dpkg-query -L iptables | egrep "xtables-(legacy|nft)-multi$"
/usr/sbin/xtables-legacy-multi
/usr/sbin/xtables-nft-multi
```

Those do not exist in iptables v1.6.0, which stretch uses. This wrapper script:

https://github.com/kubernetes/release/blob/master/images/build/debian-iptables/stretch/iptables-wrapper

Assumes that those symlinks exists and tries to update alternatives, which fails.

**Comment by wespanther** [comment]
Pulled it down, managed to get it building by changing Makefile to reference stretch in the old gcr.io/k8s-staging-build-image project and deleting the new update-alternatives additions in the dockerfile.

Before:
```
wpanther@wpanther src/k8s.io/release/images/build/debian-iptables (iptables-test *) » docker run -it gcr.io/k8s-staging-build-image/debian-iptables-amd64:stretch-v1.2.0
# which iptables
/usr/sbin/iptables
# iptables
update-alternatives: error: alternative /usr/sbin/iptables-legacy for iptables not registered; not setting
```

After:
```
wpanther@wpanther src/k8s.io/release/images/build/debian-iptables (iptables-test *) » make build ARCH=amd64
cp -r ./stretch /tmp/tmp.8bLr04lQlg/
cd /tmp/tmp.8bLr04lQlg/stretch && sed -i "s|BASEIMAGE|gcr.io/k8s-staging-build-image/debian-base-amd64:stretch-v1.1.0|g" Dockerfile
docker build \
	--pull \
	-t "gcr.io/k8s-staging-build-image"/debian-iptables-amd64:stretch-v1.2.0 \
	-t "gcr.io/k8s-staging-build-image"/debian-iptables-amd64:v0.4.2-22-g1aeff4e-dirty-stretch \
	-t "gcr.io/k8s-staging-build-image"/debian-iptables-amd64:latest-stretch \
	--build-arg=IPTABLES_VERSION=1.8.5 \
	/tmp/tmp.8bLr04lQlg/stretch

Sending build context to Docker daemon   5.12kB
Step 1/2 : FROM gcr.io/k8s-staging-build-image/debian-base-amd64:stretch-v1.1.0
stretch-v1.1.0: Pulling from k8s-staging-build-image/debian-base-amd64
Digest: sha256:bb94068a9a1f9345aa4257206a37b6343520330c2c387af74c9684fafa4161dd
Status: Image is up to date for gcr.io/k8s-staging-build-image/debian-base-amd64:stretch-v1.1.0
 ---> db72db90f398
Step 2/2 : RUN clean-install     conntrack     ebtables     ipset     iptables     kmod     netbase
 ---> Using cache
 ---> f4f95ce38a00
[Warning] One or more build-args [IPTABLES_VERSION] were not consumed
Successfully built f4f95ce38a00
Successfully tagged gcr.io/k8s-staging-build-image/debian-iptables-amd64:stretch-v1.2.0
Successfully tagged gcr.io/k8s-staging-build-image/debian-iptables-amd64:v0.4.2-22-g1aeff4e-dirty-stretch
Successfully tagged gcr.io/k8s-staging-build-image/debian-iptables-amd64:latest-stretch
wpanther@wpanther src/k8s.io/release/images/build/debian-iptables (iptables-test *) »
wpanther@wpanther src/k8s.io/release/images/build/debian-iptables (iptables-test *) » docker run -it gcr.io/k8s-staging-build-image/debian-iptables-amd64:stretch-v1.2.0
# iptables
iptables v1.6.0: no command specified
Try `iptables -h' or 'iptables --help' for more information.
```

**Comment by wespanther** [comment]
So now, all we have to do is:

1. Get the debian-base:stretch* images (at least v1.1.0) into its new home in k8s.gcr.io/build-image
2. Delete the update-alternatives parts from the stretch dockerfile
3. Build and upload the new image from that

**Comment by wespanther** [comment]
Looks like stretch-1.1.0 is the same as v1.1.0. I can go ahead and use that one.

**Comment by wespanther** [comment]
Just to be explicit, here's what is working

```
wpanther@wpanther wpanther/kubernetes/release/images/build/debian-iptables (iptables-fix !*) » make build IMAGE_VERSION=stretch-v1.1.0 DEBIAN_BASE_VERSION=stretch-v1.1.0 BASE_REGISTRY=gcr.io/k8s-staging-build-image CONFIG=stretch
cp -r ./stretch /tmp/tmp.QppI4mOzeD/
cd /tmp/tmp.QppI4mOzeD/stretch && sed -i "s|BASEIMAGE|gcr.io/k8s-staging-build-image/debian-base-amd64:stretch-v1.1.0|g" Dockerfile
docker build \
	--pull \
	-t "gcr.io/k8s-staging-build-image"/debian-iptables-amd64:stretch-v1.1.0 \
	-t "gcr.io/k8s-staging-build-image"/debian-iptables-amd64:v0.4.2-26-gbc552db-dirty-stretch \
	-t "gcr.io/k8s-staging-build-image"/debian-iptables-amd64:latest-stretch \
	--build-arg=IPTABLES_VERSION=1.8.5 \
	/tmp/tmp.QppI4mOzeD/stretch
Sending build context to Docker daemon   2.56kB
Step 1/2 : FROM gcr.io/k8s-staging-build-image/debian-base-amd64:stretch-v1.1.0
stretch-v1.1.0: Pulling from k8s-staging-build-image/debian-base-amd64
Digest: sha256:bb94068a9a1f9345aa4257206a37b6343520330c2c387af74c9684fafa4161dd
Status: Image is up to date for gcr.io/k8s-staging-build-image/debian-base-amd64:stretch-v1.1.0
 ---> db72db90f398
Step 2/2 : RUN clean-install     conntrack     ebtables     ipset     iptables     kmod     netbase
 ---> Using cache
 ---> f4f95ce38a00
[Warning] One or more build-args [IPTABLES_VERSION] were not consumed
Successfully built f4f95ce38a00
Successfully tagged gcr.io/k8s-staging-build-image/debian-iptables-amd64:stretch-v1.1.0
Successfully tagged gcr.io/k8s-staging-build-image/debian-iptables-amd64:v0.4.2-26-gbc552db-dirty-stretch
wpanther@wpanther wpanther/kubernetes/release/images/build/debian-iptables (iptables-fix !*) » docker run -it gcr.io/k8s-staging-build-image/debian-iptables-amd64:stretch-v1.1.0 iptables
iptables v1.6.0: no command specified
Try `iptables -h' or 'iptables --help' for more information.
```

**Comment by wespanther** [comment]
Coming back to this, I guess we can close it.

**Comment by justaugustus** [comment]
Yep, this should've been fixed in the last stretch images (`stretch-v1.1.0`).
For future readers, Debian base images using stretch are now deprecated: https://github.com/kubernetes/release/pull/1872

/remove-lifecycle stale
/close
